### PR TITLE
[NFC][SYCL] Break include's cycle

### PR DESCRIPTION
**Pull request by s-kanaev**
This changes is required for intel/llvm#1747.

Suppose we add `#include <CL/sycl/accessor.hpp>` in `interop_handle.hpp` . Furthermore we will add `#include <CL/sycl/interop_handle.hpp>` to `include/CL/sycl/handler.hpp`. Then, we will have the following cycle (here `->` says `which includes`):

```
handler.hpp -> interop_handle.hpp -> accessor.hpp -> handler.hpp (cycle here)
```

Also, this patch eliminates use of `reinterpret_cast` (either as it is or C-style) when converting `accessor` to `AccessorBaseHost` at host.

### CODE REVIEW

File: sycl/CMakeLists.txt [hunk 1]
```diff
@@ -11,7 +11,7 @@ option(SYCL_ADD_DEV_VERSION_POSTFIX "Adds -V postfix to version string" ON)
 set(SYCL_MAJOR_VERSION 1)
 set(SYCL_MINOR_VERSION 0)
 set(SYCL_PATCH_VERSION 0)
-set(SYCL_DEV_ABI_VERSION 1)
+set(SYCL_DEV_ABI_VERSION 2)
```
Comment (alexbatashev) [inline sycl/CMakeLists.txt:14]: From https://github.com/intel/llvm/blob/sycl/sycl/doc/ABIPolicyGuide.md:
> The development version of the library has a postfix -V that indicates breaking changes between releases.

@s-kanaev you didn't break ABI. Adding a new exported symbol is backwards-compatible change.

File: sycl/include/CL/sycl/detail/cg_types.hpp [hunk 2]
```diff
@@ -0,0 +1,307 @@
+#pragma once
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/cg_types.hpp:1]: The file is missing a license.

File: sycl/include/CL/sycl/detail/handler_proxy.hpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#pragma once
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/handler_proxy.hpp:1]: Same here.

File: sycl/source/detail/handler_proxy.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#include <CL/sycl/detail/handler_proxy.hpp>
```
Comment (alexbatashev) [inline sycl/source/detail/handler_proxy.cpp:1]: And here.

File: sycl/include/CL/sycl/detail/handler_proxy.hpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+class handler;
+
+namespace detail {
+
+class AccessorBaseHost;
+
+__SYCL_EXPORT void associateWithHandler(handler &, AccessorBaseHost *,
```
Comment (v-klochkov) [inline sycl/include/CL/sycl/detail/handler_proxy.hpp:15]: Just a question: does this function need __SYCL_EXPORT? wouldn't it be better to make it DLL_LOCAL?

**Comment by v-klochkov** [comment]
Overall Looks Good To Me. I'll mark approved (as reduction.hpp owner) after additional fixes for Alexander's comments.
Please also respond to my question regarding __SYCL_EXPORT.

File: sycl/include/CL/sycl/detail/handler_proxy.hpp (hunk 5 above)
Comment (s-kanaev) [reply sycl/include/CL/sycl/detail/handler_proxy.hpp:15]: This function is referred to from `accessor`'s constructor, which is template. Hence, the user's binary will have a reference to it.

Comment (v-klochkov) [reply sycl/include/CL/sycl/detail/handler_proxy.hpp:23]: Interesting.
@alexbatashev please take a look at this case.
IMO, it seems pretty similar to what you did in your PR here: https://github.com/intel/llvm/pull/1768
For example, handleHostData() was marked with DLL_LOCAL.
handleHostData() (sycl_mem_obj_t.hpp ) is called from buffer_impl constructor (buffer_impl.hpp), called from buffer constructor.

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/handler_proxy.hpp:23]: @v-klochkov that's different. In my case all functions are in header files. So, even if they have private linkage, they are *always* available to user. There could be another solution: move all that stuff to cpp files to make sure functions are always exported. But that would prevent compiler from optimizing most of the calls (that is especially sad for one-liners).

In Sergey's case, the code is in the *.cpp file, so he has to export the function. I believe, there's no way to break the dependency without moving some stuff to the library. Am I right, @s-kanaev?

Comment (s-kanaev) [reply sycl/include/CL/sycl/detail/handler_proxy.hpp:23]: > I believe, there's no way to break the dependency without moving some stuff to the library.

That's a precise description of my approach.

Comment (v-klochkov) [reply sycl/include/CL/sycl/detail/handler_proxy.hpp:23]: Right. I forgot that the function was in handler_proxy.cpp (I wrongly thought it was in handler_proxy.hpp).
Thank you.

**Comment by s-kanaev** [comment]
@turinevgeny @romanovvlad  any comments from you?

**Comment by romanovvlad** [comment]
> 
> 
> @turinevgeny @romanovvlad any comments from you?

No, the patch is OK with me.

**Pull request merged**